Apps/FlightSoftware/Watchdog2/src/watchdog.c

```diff
@@ -0,0 +1,355 @@
+        telbuf[4] = (uint8_t)(adc_values[ADC_BATT_LEVEL_IDX]);
+        telbuf[5] = (uint8_t)(adc_values[ADC_BATT_LEVEL_IDX] >> 8);
+        // [6,7] won't be able to read lander power in mission mode
+        // [8] currently don't read battery temp in mission mode
+        //telbuf[9] = (unit8_t)( sys_status (??) )
+        telbuf[10] = (uint8_t)(raw_battery_charge[0]);
+        telbuf[11] = (uint8_t)(raw_battery_charge[1]);
+        telbuf[12] = (uint8_t)(raw_battery_current[0]);
+        telbuf[13] = (uint8_t)(raw_battery_current[1]);
+        telbuf[14] = (uint8_t)(raw_battery_voltage[0]);
+        telbuf[15] = (uint8_t)(raw_battery_voltage[1]);
+//                telbuf[9] = (uint8_t)(adc_values[ADC_TEMP_IDX] >> 4);
+    }
+    uart0_tx_nonblocking(16, telbuf);
+}
+
+/**
+ * Function called whenever the watchdog needs to handle a byte from the hercules
+ *
+ * Assumes checksum valid.
+ */
+void watchdog_handle_hercules() {
+    // valid hercules message received
+    watchdog_flags |= WDFLAG_HERCULES_KICK;
+
+    if (uart0_rx_len != 0) {
+        /* udp packet attached */
+
+        /* echo back watchdog command header only */
+        uart0_tx_nonblocking(8, uart0_rx_header); // @suppress("Invalid arguments")
+
+        // Send telemetry data back to Hercules
+        watchdog_send_hercules_telem();
+
+        /* add this packet to the IP/UDP stack send buffer */
+        ipudp_send_packet(uart0rx.buf, uart0_rx_len); // @suppress("Invalid arguments")
+    } else {
+        /* handle watchdog reset command */
+        handle_watchdog_reset_cmd(uart0_rx_header[6]);
+        /* check if we have UDP data to send back to hercules */
+        if (hercbuf.used > 0) {
+            /* write out our buffer */
+            /* update header length first */
+            uart0_rx_header[4] = hercbuf.used & 0xFF;
+            uart0_rx_header[5] = (hercbuf.used >> 8) & 0xFF;
+            /* recompute parity */
+            uint8_t parity = 0xDC; /* sum of 0x21, 0xB0, and 0x0B */
+            /* skip parity byte (i + 3) in summation */
+            parity += uart0_rx_header[4] + uart0_rx_header[5];
+            parity += uart0_rx_header[6] + uart0_rx_header[7];
+            /* bitwise NOT to compute parity */
+            parity = ~parity;
+            /* write out new parity byte */
+            uart0_rx_header[3] = parity;
+            /* echo back watchdog command header updated version */
+            uart0_tx_nonblocking(8, uart0_rx_header); // @suppress("Invalid arguments")
+            /* note we skip sending back the telemetry here */
+            /* send the udp packet */
+            uart0_tx_nonblocking(hercbuf.used, hercbuf.buf); // @suppress("Invalid arguments")
+            /* clear hercules send buffer */
+            hercbuf.used = 0;
+            hercbuf.idx = 0;
+        } else {
+            /* echo back watchdog command header */
+            uart0_tx_nonblocking(8, uart0_rx_header); // @suppress("Invalid arguments")
```

Comment:
Reusing the volatile buffer used to receive uart data as a tx buffer seems like a recipe for disaster. It also breaks encapsulation of uart data handling, and makes this watchdog code dependent on the implementation of the uart handling code. One way to implement this a bit more safely would be to have variables defined in main() or some other parent function, and then pass those variables into this function and into the uart handling code. Have the uart code populate them with the received values, and then this code can consume them. That would remove the tight coupling that currently exists.

To be clear, my concern with this code isn't just from the perspective of code cleanliness and removing globals. My concern is primarily about race conditions in reusing these variables. I'm not seeing anything that convinces me that their use here is safe.

Reply:
I'm pretty sure there is no race condition here, as the `uart0_rx_header[]` values will only be modified by the ISR when the header has been fully processed (they will be discarded if more data comes in). Since this code block is before the "header processed OK" flag is set, the `uart0_rx_header[]` values won't be modified.

Reply:
I agree that this is a terrible hack, though. Increased encapsulation is generally probably a good idea. I originally tried to force myself to do encapsulation by only putting functions/globals intended for external use in the headers, but as you can probably see from the code, I have a LOT of random extern's and function declarations littered around the code from when I just gave up on this and made a fast, hacky, solution.